dialogue-okhttp-client/src/main/java/com/palantir/dialogue/OkHttpChannels.java

```diff
@@ -0,0 +1,161 @@
+import okhttp3.ConnectionPool;
+import okhttp3.ConnectionSpec;
+import okhttp3.Credentials;
+import okhttp3.Dispatcher;
+import okhttp3.OkHttpClient;
+import okhttp3.Protocol;
+import okhttp3.TlsVersion;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public final class OkHttpChannels {
+
+    private static final Logger log = LoggerFactory.getLogger(OkHttpChannels.class);
+    private static final boolean DEFAULT_ENABLE_HTTP2 = true;
+
+    private static final ThreadFactory executionThreads = new ThreadFactoryBuilder()
+            .setUncaughtExceptionHandler((thread, uncaughtException) -> log.error(
+                    "An exception was uncaught in an execution thread. "
+                            + "This likely left a thread blocked, and is as such a serious bug "
+                            + "which requires debugging.",
+                    uncaughtException))
+            .setNameFormat("remoting-okhttp-dispatcher-%d")
+            // This diverges from the OkHttp default value, allowing the JVM to cleanly exit
+            // while idle dispatcher threads are still alive.
+            .setDaemon(true)
+            .build();
+
+    /**
+     * The {@link ExecutorService} used for the {@link Dispatcher}s of all OkHttp clients created through this class.
+     * Similar to OkHttp's default, but with two modifications:
+     *
+     * <ol>
+     *   <li>A logging uncaught exception handler
+     *   <li>Daemon threads: active request will not block JVM shutdown <b>unless</b> another non-daemon thread blocks
+     *       waiting for the result. Most of our usage falls into this category. This allows JVM shutdown to occur
+     *       cleanly without waiting a full minute after the last request completes.
+     * </ol>
+     */
+    private static final ExecutorService executionExecutor = Executors.newCachedThreadPool(executionThreads);
+
+    /** Shared dispatcher with static executor service. */
+    private static final Dispatcher dispatcher;
+
+    static {
+        dispatcher = new Dispatcher(executionExecutor);
+        // Restricting concurrency is done elsewhere in ConcurrencyLimiters.
+        dispatcher.setMaxRequests(Integer.MAX_VALUE);
+        // Must be less than maxRequests so a single slow host does not block all requests
+        dispatcher.setMaxRequestsPerHost(256);
+    }
+
+    /** Shared connection pool. */
+    private static final ConnectionPool connectionPool = new ConnectionPool(
+            1000,
+            // Most servers use a one minute keepalive for idle connections, by using a shorter keepalive on
+            // clients we can avoid race conditions where the attempts to reuse a connection as the server
+            // closes it, resulting in unnecessary I/O exceptions and retrial.
+            55,
+            TimeUnit.SECONDS);
+
+    private OkHttpChannels() {}
+
+    public static Channel create(ClientConfiguration config, UserAgent baseAgent, TaggedMetricRegistry metrics) {
```

Comment:
so I know there a crap ton of options in a ClientConfiguration, but we dont' support all of them in this method. Could we do a Preconditions.checkState to verify the ones we don't support are empty / set to our expected defaults?

Reply:
All shall we just start with a builder straight away here? The overloads in c-j-r break my heart.

Reply:
I think the uniformity between our `XChannels.create` methods is worth something, I'd prefer to define args in dialogue-core and update them all in lock-step.

Reply:
> Could we do a Preconditions.checkState to verify the ones we don't support are empty / set to our expected defaults?

Do we want to throw if nodeSelectionStrategy is not ROUND_ROBIN? I can't imagine anyone has that set.

Reply:
There's a few users of it actually: https://internal-github/search?q=org%3Afoundry+ROUND_ROBIN&type=Code

Could do a log.warn otherwise?

Reply:
Updated.